### DISCUSSION TITLE
Return types should be list rather than Sequence?

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#9138] [Ideas]
I see the return type of `.all()` is annotated as `Sequence`, but on inspection appears to return a standard `list`:
https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult.all

Is there a deliberate reason for this? Generally, it's best to type return types with the most precise type possible.

Seems there are several others methods typed this way too.

**Comment by CaselIT** [comment]
hi,

> Is there a deliberate reason for this? Generally, it's best to type return types with the most precise type possible.

This way the public api of `all()` only guarantee that the returned value has sequence like behaviour, and it leaves future updates or extensions the ability to return different implementations.

In general I disagree that it's best to type the returns as precise as possible in a public api, since that would prevent any kind of future type change. I think a public api should return the most general type that satisfies the expected contract, that in this case is sequence like object

**Reply by Dreamsorcerer** [reply]
That's fine, just checking it was a deliberate decision.

The drawback is that any code that's expecting a list, or wants to use a list-specific method would need to pass the result through `list()`, essentially duplicating the list, if they want to be type safe. In other words it provides less flexibility to the calling code as a more general return type limits what it can do with the result (in opposition to using a general type as a parameter, which allows more flexibility to the calling code), hence why return types are usually more precise.

But, if there's the possibility you may change the type in a future version, then calling `list()` may be the safest approach anyway.

**Reply by CaselIT** [reply]
Well since modifying the list returned by `all` is not something that's common, I think in basically all cases user code could be typed to take a sequence.

**Reply by zzzeek** [reply]
> That's fine, just checking it was a deliberate decision.
> 
> The drawback is that any code that's expecting a list, or wants to use a list-specific method would need to pass the result through `list()`, essentially duplicating the list, if they want to be type safe. In other words it provides less flexibility to the calling code as a more general return type limits what it can do with the result (in opposition to using a general type as a parameter, which allows more flexibility to the calling code), hence why return types are usually more precise.
> 
> But, if there's the possibility you may change the type in a future version, then calling `list()` may be the safest approach anyway.

I think it goes beyond that; in the very early days of SQLAlchemy, I was in a very bad habit of "OK I have no plans to change this so it's public" for just about everything, and that was an enormous mistake.  even today there's lots of "seemingly public" things that should not be public.       over many years I learned the best approach for any code you write should be, "everything I write starts as completely private / underscore API.   Then carve out those tiny little bits of it that are the "public API" - and for that part of it, I have to write three pages of docs and a dozen unit tests".    That is, always keep the public API surface as small as possible, regardless of whether or not "it might change".  The decision making goes in the other direction; write as much of the code as possible as private, with at most the possibility that "maybe this will be public someday".

So in this case, we have zero intention of any result methods of implying mutability, and that the object is actually a plain list is strictly an implementation detail.    Also indeed if people started mutating that list and then complaining that something about that weren't working, we'd have to go in and say "OK right this was not meant to be mutable and if it's misleading people we will have to change this to be an immutable view".   That's happened a bunch, and is why you will see [ReadOnlyColumnCollection](https://github.com/sqlalchemy/sqlalchemy/blob/2f91dd79310657814ad28b6ef64f91fff7a007c9/lib/sqlalchemy/sql/base.py#L2042) and [MetaData.tables is a 'FacadeDict'](https://github.com/sqlalchemy/sqlalchemy/blob/2f91dd79310657814ad28b6ef64f91fff7a007c9/lib/sqlalchemy/sql/schema.py#L5212)  - people started mutating those collections directly and complained that things did not work after that.   I'm sort of hoping that doesn't happen with `Result.all()` etc. as plain `list` is going to be more performant than a custom extension type but now I'm a little worried.

**Reply by Dreamsorcerer** [reply]
> I'm sort of hoping that doesn't happen with `Result.all()` etc. as plain `list` is going to be more performant than a custom extension type but now I'm a little worried.

Personally, I just had an annotation as `list`, which resulted in a type error once I upgraded to v2. So, I merely needed to update the annotation, I wasn't actually trying to use any list-specific things, just thought it was odd to be a `Sequence`.

**Comment by zzzeek** [comment]
I agree with @CaselIT  here, if the thing we returned were a `_SuperFastListLikeThing`, we'd type that as `Sequence`, not the implementation type.   It's also not expected to be mutable.    Someday it could be some kind of a list-like view of some other internal representation.

**Comment by AlexanderPodorov** [comment]
Just want to leave a comment about that. I agree with @CaselIT about typing of returns, especially if something is expected to be immutable, then `Sequence` would be a good choice.
One nice thing that `list` supports, but `Sequence` does not, is `+` operator, so in order to concatenate two sequences we will need to use `itertools.chain`, `list(...)` or something else. If `Sequence` could support `+` one day, that would be really nice.

**Reply by zzzeek** [reply]
that's a good point, i wonder why Sequence doesn't support + ?

**Reply by zzzeek** [reply]
hmm MutableSequence?   not really ready to do that here

**Reply by Dreamsorcerer** [reply]
Well, it'd fail at runtime for one:
```
>>> [] + ()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only concatenate list (not "tuple") to list
```

A `TypeVar` bounded by `Sequence` could be safe potentially, but not any arbitrary `Sequence` implementation...

**Reply by Dreamsorcerer** [reply]
Even `MutableSequence` only appears to support `+=`, which essentially should work given any iterable:
https://github.com/python/typeshed/blob/main/stdlib/typing.pyi#L504

**Reply by AlexanderPodorov** [reply]
@Dreamsorcerer , good find. Although `tuple` is also `Sequence`, it's weird to use it as `Sequence`. E.g. we can't do `Sequence[int, bool, str]` and have to use `tuple[int, bool, str]` explicitly. Anyway it looks like it's some quirks of existing implementation.

**Reply by Dreamsorcerer** [reply]
If you're looking for a sequence, it's not weird to use a tuple (i.e. `tuple[T, ...] == Sequence[T]`). It'd be weird if you were looking for a fixed number of elements and then used a sequence or list (i.e. `tuple[int, str] != Sequence[int | str]`).

The main point is that if sqlalchemy were to change the return type in future, it'd likely break any `+` operations (unless both sides of the operation came from sqlalchemy functions).

If you wanted to commit to being able to concatenate lists in all future versions of sqlalchemy, then a custom return type that is a little more specific (e.g. inherit from Sequence and add an `__add__()`) could allow that.

**Reply by AlexanderPodorov** [reply]
Yes, I agree to all what you said, you're totally right. I didn't mean to change anything, just mentioned that `+` would be nice :-)
I just wanted to share my personal opinion that not all types of `tuples`s can be represented as `Sequence`.
Like `tuple[int, bool]` can't be represented as `Sequence[T]`, no matter what `T` is. It looks weird to me that `tuple` inherits from `Sequence`, but can't be correctly typed in all cases with just `Sequence`. But that's how stdlib is implemented ;-)

**Reply by Dreamsorcerer** [reply]
Well, tuple is a more precise type, so I'm not sure it's that weird. Could be nice to have a generic `FixedSequence` or something though, but it's not relevant to the methods we are talking about here.

If sqlalchemy wants to explicitly support `+` then a custom type should be possible. The decision would need to be made as to whether it should always support concatenating to `list`, or should just support concatenating to the same type.

e.g. Depending on the choice, you could do roughly one of these:
```
class SASequence(Sequence[T]):
    def __add__(self, other: SASequence[T] | list[T]): ...

class SASequence(Sequence[T]):
    def __add__(self, other: SASequence[T]): ...
```

But, then you might end up needing a `cast()` or type factory or something when creating the `list`s.

**Reply by CaselIT** [reply]
Curious what's the use case for using + in this particular case?

**Reply by AlexanderPodorov** [reply]
@Dreamsorcerer : yes it could be one of the options. But `Sequence` is fine for now, whatever we can, we can do on our end.
@CaselIT : nothing related to sqlalchemy really, just my particular use case.
I think we can finish this discussion, anything else is just off topic. Thank you guys!

**Comment by XCanG** [comment]
I see this as a potential issue as for now. I think it is better to change only when an actual changes to the code will be made.

The thing is, as for now it leave you into two options:

* Silence error with `# type: ignore[override]` but it may produce errors in the future. And adding that to like 100+ functions in API is bad.
* Cast `list()` on result and get performance lost. If you use something like `timeit` to check how many time will take to just get list and get `list(my_list)` you will see about 100x times overhead (for me it 113.44x)

And in the code as of right now it's `List` type for `.all()`.
https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/engine/result.py#L537-L560

The better way would be use `List` type now. When something new being added, change types and SQLAlchemy version, ex from 2.0.* to 2.1.0

**Reply by CaselIT** [reply]
I'm not sure what's the issue is with Sequence. What list method are you looking for that isn't in Sequence?

**Reply by XCanG** [reply]
I have errors from type checker, I'm working with Fastapi and Pydantic models. For example take a look at simple code:
```py
async def get_all(session: AsyncSession, limit: int, offset: int, order_by: list[str]) -> list[User]:
    result = await session.execute(select(User).order_by(*order_by).limit(limit).offset(offset))
    return result.scalars().all()
```
With that I get error because I get `Sequence` instead of `List`

**Reply by CaselIT** [reply]
Can't you type it as `-> Sequence[User]:`? What's the issue with Sequence?

**Reply by XCanG** [reply]
Because later I load it into Pydantic model and it should be type of `list`

**Reply by XCanG** [reply]
I get message like:
```
Argument of type "Sequence[User]" cannot be assigned to parameter "data" of type "list[User]" in function "__init__"
  "Sequence[User]" is incompatible with "list[User]"
```

**Reply by XCanG** [reply]
And I have like hundreds of errors of similar type for any `.any()` function.

**Reply by zzzeek** [reply]
explicit is better than implicit, you should use list() or a list comprehension.   Result is not guaranteed to return a list.

**Reply by XCanG** [reply]
@zzzeek But that will take 100x overhead on almost all endpoints without any reason, it's worst case scenario to loose performance when there is no need for it.

**Reply by CaselIT** [reply]
> Because later I load it into Pydantic model and it should be type of `list`

Can't you make that of type Sequence?

> @zzzeek But that will take 100x overhead on almost all endpoints without any reason, it's worst case scenario to loose performance when there is no need for it.

You can just use cast or type ignore.

**Reply by zzzeek** [reply]
100x is a complete exaggeration, the extra list() or cast() call would have overhead very negligible compared to everything else going on with constructing objects and fetching database results.    run some cProfile results within real-world scenarios and you'll see additional cpu time spent a list() or cast() call in full context for these operations is tiny compared to the overall burden of the rest of the program.

**Reply by XCanG** [reply]
> Can't you make that of type Sequence?
> You can just use cast or type ignore.

I explained 2nd in my original post, that would be bad in the future + there would need to place in in ~100 functions, which is bad on it's own.

About the first I can't because it need to be JSON serializable when it auto-convert it later for web response.

**Reply by zzzeek** [reply]
if this is really that much of an issue, I will accept a PR for a new method ``as_list()`` to all result objects that always returns a List that's guaranteed to be mutable.

**Reply by zzzeek** [reply]
I'm just personally not a big fan of `x = result.all();  now_mutate_the_list(x)`.   `result` is oriented around read-only operations.

**Reply by zzzeek** [reply]
see for example https://peps.python.org/pep-0249/#fetchall

**Reply by XCanG** [reply]
@zzzeek ok, I may not say it clearly, but comparison done not to the rest of the code, but to the getting **list of variable** vs **casting list on list variable** (so that no changing type was changed, but function is still evaluated)

**Reply by XCanG** [reply]
> `result` is oriented around read-only operations.

I'm not disagree with this statement, but my statement for solely type matching and not throwing errors, because as for now type of Sequence adding a lot of error in the project while type underneath was really the same (list: https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/engine/result.py#L537-L560)

**Reply by XCanG** [reply]
> see for example https://peps.python.org/pep-0249/#fetchall

I don't see a problem of typing it as `list[tuple]` rather than `Sequence[Sequence]`

**Reply by zzzeek** [reply]
Compromise is being offered , which AFAICT would solve your problem.

**Reply by CaselIT** [reply]
> > see for example https://peps.python.org/pep-0249/#fetchall
> 
> I don't see a problem of typing it as `list[tuple]` rather than `Sequence[Sequence]`

And I don't see the problem on using cast... This is getting unreasonable, sqlalchemy is not used only by your project..

> About the first I can't because it need to be JSON serializable when it auto-convert it later for web response.

I don't understand what's the issue with sequence and json. The serializer uses the actual type that's a list, so I see no problem here.


> if this is really that much of an issue, I will accept a PR for a new method `as_list()` to all result objects that always returns a List that's guaranteed to be mutable.

How about allowing the container to be passed in the `all()` function? This was we don't have this drama the next time someone insists on wanting a tuple.

**Reply by zzzeek** [reply]
or just tell people to call `list(result)`, that would after all be the most pythonic way to do this.

**Reply by CaselIT** [reply]
that's indeed probably the best option here. That uses an iterator and does not create an intermediate list

**Reply by XCanG** [reply]
> I don't understand what's the issue with sequence and json. The serializer uses the actual type that's a list, so I see no problem here.

The thing is as of moment of writing even with errors in console code works, it just that type checker throw errors in console.

If you do something stupid in python like:
```py
x: int = "temp"
y: list = "5"
print(x + y)
```
this code will work when being runned, but type checker will argue about types. I have exactly the same situation as I have error with Sequence type, but code is still works because list is what it actually stored from. But it will require to rewrite type with ignore, cast type or changing all types in the models. As much as I like other changes I really like how it was been typed in 1.4 with stubs.

**Reply by Dreamsorcerer** [reply]
> I really like how it was been typed in 1.4 with stubs.

i.e. Not typed? Just disable your type checkers if you don't want it to be typed.
My main reason for upgrading to sqlalchemy 2 quickly is because, even with the stubs package, most code ended up untyped in 1.4.

**Comment by AlexanderPodorov** [comment]
@XCanG , as mentioned above, you can use `Sequence[T]` in `pydantic` model:
```python
class Model(BaseModel):
    items: Sequence[Item]
```
That's what I always do.
I'm curious about how you're not having issues with type checker while assigning `sqlalchemy` objects directly into `pydantic` models.
Technically `sqlalchemy` object have to be parsed into `pydantic` model before returning a response (in the very beginning I did the same thing, but switched to parsing to pydantic model after having issues since certain `fastapi` release...).

**Reply by CaselIT** [reply]
>  Technically `sqlalchemy` object have to be parsed into `pydantic` model before returning a response (in the very beginning I did the same thing, but switched to parsing to pydantic model after having issues since certain `fastapi` release...).

That's indeed the case. Pydantic will create new objects for each sqlalchemy instance, validate each property and place them in a new list, but apparently using `cast` is a performance problem..

**Reply by XCanG** [reply]
Technically there is no issue there now, code is work even with errors, it just that previously type checker has `Unknown` type, so it wasn't arguing about incorrect types and now when it typed and type mismatch it sees it as an issue. While code is working, IDE shows a lot of errors which is a nightmare to look at (you can't probably ignore hundred of errors in your console).

**Reply by XCanG** [reply]
I think I may stick with this idea as it best of 3. Because I don't like loosing performance with cast and I don't like add `# type: ignore` and potentially hiding bugs in the future.

**Reply by CaselIT** [reply]
note that doing `list(result)` should not cause any performance issue, and return a list

**Comment by Suaralanre** [comment]
I solved the problem with python generics:

```
 from typing import TypeVar, Generic, Sequence

 T = TypeVar("T")

 Class MyPydanticSchema(BaseModel, Generic[T]):

    x: Sequence[T]
    
    ---------------------------------------------------------------
    stmt = select(Table).where(a == b)
    data = db.scalars(stmt).all()
    # type data = Sequence[Table]
    
    --------------------------------------------------------------
    print(MyPydanticSchema(x=data)
```
